src/workerd/api/actor.h

```diff
@@ -13,27 +13,95 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
+public:
+  DurableObjectRpcStub(
+      IoOwn<OutgoingFactory> outgoingFactory,
+      RequiresHostAndProtocol requiresHost,
+      IsInHouse inHouse,
+      CompatibilityFlags::Reader flags)
+    : Fetcher(kj::mv(outgoingFactory), requiresHost, (inHouse == YES)), flags(flags) {}
+
+  bool isRegisteredName(kj::StringPtr name) {
+    // We don't need to check `fetch` or `constructor` because `fetch` is known at compile time and
+    // `constructor` is a JS keyword; neither uses name interception. This needs to be updated
+    // whenever a new "registered name", aka handler, is added to Durable Objects.
+    if (name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  kj::Maybe<jsg::JsValue> getNamed(jsg::Lock& js, kj::StringPtr name) override {
+    if (flags.getJsRpcStub()) {
+      JSG_REQUIRE(!isRegisteredName(name), Error, "cannot call registered method '", name, "'");
+
+      // Named intercept is enabled, this means we won't default to legacy behavior.
+      // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
+      //             this is just a temporary placeholder for testing.
+      return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
+          [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+            return js.error("JsRpcStub is unimplemented");
+          }
+      ));
+    }
+    // JS RPC has not been enabled, default to legacy behavior!
+    return kj::none;
+  }
+
+  // WARNING: Adding a new method to the Durable Object stub can conflict with RPC method names
```

Comment:
This warning applies to adding methods to `Fetcher` too, right? Since it is inherited?

Reply:
Yeah, it will apply to any type that uses `JSG_NAMED_INTERCEPT`. But as mentioned in the other comment, we can tweak how this works by changing the flags we use to register the intercept callbacks. The current choice was made in an attempt to make it possible for existing jsg objects to continue to use the other macros easily (JSG_METHOD, JSG_*_PROPERTY, etc) while also supporting interception. But we have the freedom to change how the intercept works right now.